### ISSUE TITLE
[java] UseDiamondOperator false-positive inside lambda

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1723]
Rule: [UseDiamondOperator](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#usediamondoperator)

PMD version is 6.12.0.

This is related to [another issue with the same rule](https://github.com/pmd/pmd/issues/1624).

This time I get it in a lambda:

```java
Runnable someAction = () -> {
    var foo = new ArrayList<String>(5);
    String x = foo.get(0);
    System.err.println(x);
};
```

By the way, thank you for working on PMD! It's an amazing tool!

**Comment by jsotuyod** [comment]
@boris-petrov that's interesting… I see nothing at first sight in the rule that should cause this scenario to fail, but I'll have to dig deeper.

Thanks for the report!

**Comment by krichter722** [comment]
In the example code

```
Runnable someAction = () -> {
	var foo = new ArrayList<String>(5);
	System.err.println(foo);
};
```

 the explicit type argument is redundant as the code compiles without the specification.

It becomes necessary as soon as the generic type is accessed, e.g. when assigning `String x = foo.get(0)`.

**Comment by adangel** [comment]
FYI: I've updated the updated the code sample, as @krichter722 correctly pointed out, that in the original example, the type parameter where indeed unnecessary (and the rule would have been accidentally correct) - this original example however looks artificial. I guess it is more common that the declared var is later on used to access (in this case) a list element. And then, the type parameter are required. Thus this is still a false positive, and this will be fixed via #2392.
